### ISSUE TITLE
PMIx runtime error.

### ISSUE THREAD

**Issue by connorourke** [open-mpi/ompi#6742]
Thank you for taking the time to submit an issue!

## Background information
I am running into problems across multiple machines, and multiple versions of openmpi when trying to get an mpi4py code that cycles through a loop, spawning and closing down multiple instances of a fortran executable, to run successfully. (see https://github.com/open-mpi/ompi/issues/6727 and https://github.com/open-mpi/ompi/issues/6710). And I was hoping some one would be able to explain the root of this error and as to whether OpenMPI is fit to handle this type of code?

### What version of Open MPI are you using? (e.g., v1.10.3, v2.1.0, git branch name and hash, etc.)

v 3.1.1

### Describe how Open MPI was installed (e.g., from a source/distribution tarball, from a git clone, from an operating system distribution package, etc.)

Centrally managed installation module on HPC system. I'm guessing installed from a tarball:

```  
Configure command line: '--prefix=/shared/ucl/apps/openmpi/3.1.1/gnu-4.9.2'
                          '--with-sge' '--without-verbs'
```

### Please describe the system on which you are running

* Operating system/version: 
Red Hat Enterprise Linux Server release 7.4
* Computer hardware: 
Intel(R) Xeon(R) CPU E5-2650 v4 @ 2.20GHz
* Network type: 
Happens on a single node

-----------------------------

## Details of the problem

When running the following code:

```python

from mpi4py import MPI
import sys
import subprocess

my_comm = MPI.COMM_WORLD
my_rank = MPI.COMM_WORLD.Get_rank()
size = my_comm.Get_size()
no_workers = size -1


def spawn():
        mpi_info = MPI.Info.Create()
        mpi_info.Set("hostfile",  MPI.Get_processor_name())
        executable = "hello"
        commspawn = MPI.COMM_SELF.Spawn(executable, args="", maxprocs=4, info=mpi_info)
        commspawn.Barrier()
        commspawn.Disconnect()
        mpi_info.Free()
        return True


if __name__ == "__main__":


    sys_excepthook = sys.excepthook
    def mpi_excepthook(type, value, traceback):
        sys_excepthook(type,value,traceback)
        MPI.COMM_WORLD.Abort(1)

    sys.excepthook = mpi_excepthook


    for i in range(10000):
        print("running",i)
        spawn()
        my_comm.Barrier()

    my_comm.Barrier()
    MPI.Finalize()
```

where the Fortran executable spawned is as follows:

```fortran
   program hello
   include 'mpif.h'
   integer::rank, size, ierror, tag, status(MPI_STATUS_SIZE)
   integer::mpi_comm_parent,length,dum_err
   CHARACTER(MPI_MAX_ERROR_STRING):: message
   CHARACTER(len=256)::hostname

       call MPI_INIT(ierror)
       call MPI_COMM_SIZE(MPI_COMM_WORLD, size, ierror)
       call MPI_COMM_RANK(MPI_COMM_WORLD, rank, ierror)
       CALL MPI_GET_PROCESSOR_NAME(HOSTNAME,hostlength, hosterror)
       print*, 'From node', rank,'on',trim(hostname),': Hello world'




        call mpi_COMM_get_parent(mpi_comm_parent,ierr)

        IF (mpi_comm_parent .ne. MPI_COMM_NULL) THEN
            print*,"calling disconnect"
            CALL MPI_BARRIER(mpi_comm_parent,ierr)
            CALL MPI_COMM_DISCONNECT(mpi_comm_parent,ierr)
        end if

   call MPI_FINALIZE(ierror)
   end
```

I get the error:

```shell
[node-k10b-032:17826] PMIX ERROR: ERROR in file gds_dstore.c at line 1178
[node-k10b-032:17826] PMIX ERROR: OUT-OF-RESOURCE in file gds_dstore.c at line 1280
[node-k10b-032:17826] PMIX ERROR: OUT-OF-RESOURCE in file gds_dstore.c at line 2331
[node-k10b-032:17826] PMIX ERROR: OUT-OF-RESOURCE in file gds_dstore.c at line 3146
[node-k10b-032:17826] PMIX ERROR: OUT-OF-RESOURCE in file gds_dstore.c at line 3178
[node-k10b-032:17826] PMIX ERROR: OUT-OF-RESOURCE in file server/pmix_server.c at line 2151
[node-k10b-032:15154] PMIX ERROR: OUT-OF-RESOURCE in file client/pmix_client.c at line 228
[node-k10b-032:15154] OPAL ERROR: Error in file pmix2x_client.c at line 109
*** An error occurred in MPI_Init
*** on a NULL communicator
*** MPI_ERRORS_ARE_FATAL (processes in this communicator will now abort,
***    and potentially your MPI job)
[node-k10b-032:15154] Local abort before MPI_INIT completed completed successfully, but am not able to aggregate error messages, and not able to guarantee that all other processes were killed!
-------------------------------------------------------
Child job 1189 terminated normally, but 1 process returned
a non-zero exit code. Per user-direction, the job has been aborted.
-------------------------------------------------------
[node-k10b-032:17826] *** Process received signal ***
[node-k10b-032:17826] Signal: Segmentation fault (11)
[node-k10b-032:17826] Signal code: Address not mapped (1)
[node-k10b-032:17826] Failing at address: 0xe0
[node-k10b-032:17826] [ 0] /lib64/libpthread.so.0(+0xf5e0)[0x2b62fb9cc5e0]
[node-k10b-032:17826] [ 1] /shared/ucl/apps/openmpi/3.1.1/gnu-4.9.2/lib/openmpi/mca_pmix_pmix2x.so(pmix_ptl_base_lost_connection+0xc5)[0x2b62ff5ed885]
[node-k10b-032:17826] [ 2] /shared/ucl/apps/openmpi/3.1.1/gnu-4.9.2/lib/openmpi/mca_pmix_pmix2x.so(pmix_ptl_base_recv_handler+0x361)[0x2b62ff5efe61]
[node-k10b-032:17826] [ 3] /shared/ucl/apps/openmpi/3.1.1/gnu-4.9.2/lib/libopen-pal.so.40(opal_libevent2022_event_base_loop+0x7a6)[0x2b62fa9feda6]
[node-k10b-032:17826] [ 4] /shared/ucl/apps/openmpi/3.1.1/gnu-4.9.2/lib/openmpi/mca_pmix_pmix2x.so(+0x884ae)[0x2b62ff5c54ae]
[node-k10b-032:17826] [ 5] /lib64/libpthread.so.0(+0x7e25)[0x2b62fb9c4e25]
[node-k10b-032:17826] [ 6] /lib64/libc.so.6(clone+0x6d)[0x2b62fbcd134d]
[node-k10b-032:17826] *** End of error message ***
/shared/ucl/apps/openmpi/ucl-wrapper-omp3+/bin/mpirun: line 28: 17826 Segmentation fault      (core dumped) $MPI_LAUNCHER -mca pml ob1 -mca btl tcp,vader,self "$@"
```

The code runs happily for an indeterminate number of steps (the step at which the error occurs varies).

**Comment by artpol84** [comment]
This kind of error met occur only if posix_fallocate fails:
https://github.com/open-mpi/ompi/blob/v3.1.x/opal/mca/pmix/pmix2x/pmix/src/mca/pshmem/mmap/pshmem_mmap.c#L79

Check your file system. I guess it should be in /tmp directory.

**Comment by artpol84** [comment]
Im taking it back: initially error appears here:
https://github.com/open-mpi/ompi/blob/v3.1.x/opal/mca/pmix/pmix2x/pmix/src/mca/gds/ds12/gds_dstore.c#L1178
So it’s not

**Comment by rhc54** [comment]
@connorourke Could you try building OMPI against the head of the PMIx v2.2 branch and see if the problem is still present?

**Comment by ggouaillardet** [comment]
that sounds like a memory leak ...

**Comment by connorourke** [comment]
Sorry for the delay. I have refactored the code that I was working on, so that it spawns the executable only once, passes the data back and forth and tears down from within.

The code runs fine on a single node, but now throws up a new error if I try to run on more than one node. 

The error is a s follows:

```
[node-k10a-004:08972] [[26813,1],24] selected pml cm, but peer [[26813,2],0] on node-k10a-004 selected pml ob1
[node-k10a-004:08972] [[26813,1],24] ORTE_ERROR_LOG: Unreachable in file dpm/dpm.c at line 433
--------------------------------------------------------------------------
ORTE does not know how to route a message to the specified daemon
located on the indicated node:

  my node:   node-k10a-015
  target node:  node-k10a-004

This is usually an internal programming error that should be
reported to the developers. In the meantime, a workaround may
be to set the MCA param routed=direct on the command line or
in your environment. We apologize for the problem.

```

I have hacked together a test code that reproduces the problem:

hello.f90:

``` fortran
      program hello
      implicit none
      include 'mpif.h'
      integer::rank, size, ierror, tag, status(MPI_STATUS_SIZE),step,i
      integer::mpi_comm_parent,length,dum_err,parent_stop,ierr,hostlength
      integer::nargs
     CHARACTER(MPI_MAX_ERROR_STRING):: message
      CHARACTER(len=256)::hostname,spawndir,arg

       call MPI_INIT(ierror)
       call MPI_COMM_SIZE(MPI_COMM_WORLD, size, ierror)
       call MPI_COMM_RANK(MPI_COMM_WORLD, rank, ierror)
       CALL MPI_GET_PROCESSOR_NAME(HOSTNAME,hostlength,ierr)
       call mpi_COMM_get_parent(mpi_comm_parent,ierr)


       nargs = command_argument_count()
       DO i=1,nargs
            call getarg(i,arg)
            SELECT CASE(arg)
            CASE('-spawndir')
                CALL getarg(i+1,spawndir)
                CALL chdir(TRIM(spawndir))
            END SELECT
       END DO

        parent_stop = 0
        step = 0
        do !i=0,9     

            CALL MPI_BCast(parent_stop, 1, MPI_INTEGER, 0, mpi_comm_parent, ierr)
            CALL MPI_BARRIER(mpi_comm_parent,ierr)

            if (parent_stop > 0) then
                if (rank==0) print*, "stop increased in child on host:",trim(hostname)," exiting"
                exit
            end if
            step = step + 1
        end do

   IF (mpi_comm_parent .ne. MPI_COMM_NULL) THEN
      CALL MPI_BARRIER(mpi_comm_parent,ierr)
      CALL MPI_COMM_DISCONNECT(mpi_comm_parent,ierr)
   end if

   call MPI_FINALIZE(ierr)
   end
```

and test.py

```python
from mpi4py import MPI
import sys
import numpy as np

my_comm = MPI.COMM_WORLD
my_rank = MPI.COMM_WORLD.Get_rank()
size = my_comm.Get_size()

if __name__ == "__main__":


    sys_excepthook = sys.excepthook
    def mpi_excepthook(type, value, traceback):
        sys_excepthook(type,value,traceback)
        MPI.COMM_WORLD.Abort(1)

    sys.excepthook = mpi_excepthook
    spawn_args=list()
    spawn_args.append('-spawndir')
    spawn_args.append(str(my_rank))

    mpi_info = MPI.Info.Create()
    mpi_info.Set("host",  MPI.Get_processor_name())
    executable = 'hello'
    stop = np.array(0,dtype='i')


    commspawn = MPI.COMM_SELF.Spawn(executable, args=spawn_args, maxprocs=4, info=mpi_info)

    for step in range(12000):
        print("running step",step,"on: ",my_rank)
        if step > 10:
           stop = np.array(1,dtype='i')
           print("ppfit stop updated",stop,type(stop),int(stop) > 0)

        commspawn.Bcast(stop, root=MPI.ROOT)
        commspawn.Barrier()
        if stop > 0:
            break
    print("PY broke from loop")
    commspawn.Barrier()
    commspawn.Disconnect()

    MPI.COMM_WORLD.Barrier()
    MPI.Finalize()
```

which i run as:

` mpirun --map-by node  -np 8 python3 ./test.py`

`test.py` spawns several instances of `hello`, and then loops passing stop to `hello` each step. When it passes a certain number of steps `stop` is increased and the loop is exited. 

This is with openmpi v3.1.1. I also tried v3.1.4 built with pmix v2.2.3, and got the following error:
```

> 
> [node-k10a-004:11857] [[24411,1],5] ORTE_ERROR_LOG: Error in file dpm/dpm.c at line 433
> [node-k10a-004:11924] [[24411,3],0] ORTE_ERROR_LOG: Not found in file dpm/dpm.c at line 433
> [node-k10a-004:11856] [[24411,1],3] ORTE_ERROR_LOG: Error in file dpm/dpm.c at line 433
> [node-k10a-004:11928] [[24411,4],0] ORTE_ERROR_LOG: Not found in file dpm/dpm.c at line 433
> [node-k10a-004:11926] [[24411,3],2] ORTE_ERROR_LOG: Not found in file dpm/dpm.c at line 433
> [node-k10a-004:11925] [[24411,3],1] ORTE_ERROR_LOG: Not found in file dpm/dpm.c at line 433
> [node-k10a-004:11927] [[24411,3],3] ORTE_ERROR_LOG: Not found in file dpm/dpm.c at line 433
> [node-k10a-004:11930] [[24411,4],2] ORTE_ERROR_LOG: Not found in file dpm/dpm.c at line 433
> --------------------------------------------------------------------------
> It looks like MPI_INIT failed for some reason; your parallel process is
> likely to abort.  There are many reasons that a parallel process can
> fail during MPI_INIT; some of which are due to configuration or environment
> problems.  This failure appears to be an internal failure; here's some
> additional information (which may only be relevant to an Open MPI
> developer):
> 
>   ompi_dpm_dyn_init() failed
>   --> Returned "Not found" (-13) instead of "Success" (0)
> --------------------------------------------------------------------------
> [node-k10a-004:11929] [[24411,4],1] ORTE_ERROR_LOG: Not found in file dpm/dpm.c at line 433
> [node-k10a-004:11931] [[24411,4],3] ORTE_ERROR_LOG: Not found in file dpm/dpm.c at line 433
> [node-k10a-004:11939] [[24411,8],0] ORTE_ERROR_LOG: Not found in file dpm/dpm.c at line 433

```

**Comment by rhc54** [comment]
This is ages old, but the problem is right here:
```
[node-k10a-004:08972] [[26813,1],24] selected pml cm, but peer [[26813,2],0] on node-k10a-004 selected pml ob1
```
The processes on the different nodes are seeing very different environments, leading them to select a different pml component. This is the component that manages communications and it cannot differ between processes.

You just need to add ```--mca pml ob1``` to your ```mpirun``` cmd line to ensure they all pick the same pml component.

**Comment by jsquyres** [comment]
Given that this issue is so old, and it looks like @rhc54 answered the question, I'm going to go ahead and close the issue.  Feel free to re-open if the same issue is persisting, or open a new issue if something different is happening.

Thanks!
